modules/core/src/test/java/org/apache/ignite/internal/processors/database/IgniteDbPutGetAbstractTest.java

```diff
@@ -271,6 +301,56 @@ public void testPutGetLarge() throws Exception {
         assertNull(cache1.get(1));
     }
 
+    /**
+     * Tests that putting a large entry, which size is greater than WAL buffer/segment size, results in CacheException.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testPutLargeEntry() throws Exception {
+        assertTrue(
+            "Primary key should correspond to the node with small wal buffer size.",
+            smallWalBufSizeNodeIdx == 0);
+
+        IgniteCache<Integer, byte[]> atomicCache = grid(0).cache("atomic");
+        Integer atomicPrimaryKey = primaryKey(atomicCache);
+
+        // New value which is greater than WAL segment size / WAL buffer.
+        byte[] newVal = new byte[DFLT_WAL_SEGMENT_SIZE / 2];
+
+        assertThrows(
+            log,
+            () -> atomicCache.put(atomicPrimaryKey, newVal),
+            CacheException.class,
+            null);
+        assertNull("Unexpected non-null value.", atomicCache.get(atomicPrimaryKey));
+        assertTrue("Unexpected system critical error.", failedNodes.isEmpty());
+
+        // Check backup scenario.
+        if (gridCount() > 1) {
+            Integer atomicBackupKey = backupKey(atomicCache);
+            Ignite primaryNode = primaryNode(atomicBackupKey, atomicCache.getName());
+
+            // Primary node should be updated successfully,
+            // however, backup node should fail because of size of the new entry does not allow to write it to WAL.
+            assertThrows(
+                log,
+                () -> atomicCache.put(atomicBackupKey, newVal),
+                CacheException.class,
+                "Failed to update keys");
+
+            assertTrue(
+                "Unexpected value.",
+                Arrays.equals((byte[])primaryNode.cache(atomicCache.getName()).get(atomicBackupKey), newVal));
+
+            assertTrue(
```

Comment:
As well as here, you could write `assertThat(a, contains(b))` or something like that

Reply:
Wow, `hamcrest` 2.x is available in Ignite codebase! Will use it.